mobius-android/src/main/java/com/spotify/mobius/android/MutableEffectLiveData.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius.android;
+
+import android.arch.lifecycle.LifecycleOwner;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.Observer;
+
+/**
+ * An internal implementation of {@link EffectLiveData} that allows posting values. This
+ * implementation is backed by an Android Live data and also uses an {@link Accumulator} to allow
+ * queueing up values being sent.<br>
+ * There are two methods to send values - post and postTransient, the first will queue values, the
+ * other will not. See method docs.
+ *
+ * @param <T> The type of data to store and queue up
+ */
+final class MutableEffectLiveData<T> implements EffectLiveData<T> {
+  private final MutableLiveData<Accumulator<T>> data = new MutableLiveData<>();
+
+  private static <T> Observer<Accumulator<T>> unfold(Observer<? super T> observer) {
+    return accumulator -> {
+      if (accumulator != null) {
+        accumulator.handle(observer::onChanged);
+      }
+    };
+  }
+
+  @Override
+  public boolean hasActiveObservers() {
+    return data.hasActiveObservers();
+  }
+
+  @Override
+  public boolean hasObservers() {
+    return data.hasObservers();
+  }
+
+  @Override
+  public void observe(LifecycleOwner owner, Observer<? super T> observer) {
+    data.observe(owner, unfold(observer));
+  }
+
+  @Override
+  public void observeForever(Observer<? super T> observer) {
+    data.observeForever(unfold(observer));
+  }
+
+  @Override
+  public void removeObservers(LifecycleOwner owner) {
+    data.removeObservers(owner);
+  }
+
+  /**
+   * This method will post the value via the live thread, behaving similarly to MutableLiveData's
+   * postValue, with the difference being that if multiple values are being sent while nothing is
+   * observing the data, they will be queued up along with any other calls made to post values.
+   *
+   * @param value The value to emit, or alternatively if nothing is observing, to queue up
+   */
+  void post(T value) {
+    synchronized (data) {
+      if (data.getValue() == null) {
+        data.postValue(new Accumulator<T>().append(value));
+      } else {
+        data.postValue(data.getValue().append(value));
```

Comment:
I guess it's not really necessary to call `data.postValue()` here. Maybe this method could be simplified to something like:

```
  synchronized(data) {
    Accumulator<T> acc = data.getValue();
    if (acc == null) {
       acc = new Accumulator<>();
       data.postValue(acc);
    }
    acc.append(value);
  }
```
A reason is that `data.postValue` is scary since it might miss things that are not posted on the main thread. In general, perhaps using a LiveData isn't the safest thing to do.

Reply:
`data.postValue()` is necessary.

First, the `getValue()` call will only be null once, when there's nothing posted at all. Afterwards, the object posted will be then received and will internally use up its values. 

When we call post for a second time, we will append new values to the object, but unless we posted it again into the MutableLiveData, whoever is observing the live data won't get notified that we've sent out new things.

Ideally (and originally) I had a more of FP approach where instead of append there was a function called Plus that always returned an new Accumulator, but since then for optimizations of allocations I've converted it to a single mutable object.

Bottom line though, we definitely need to call postValue every time, otherwise observers won't be notified by the live data that there's new things to observe